subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2356 @@
+
+	/* Copy the header into the buffer */
+	net_buf_add_mem(buf, &data->payload_header, data->payload_header.bHeaderLength);
+
+	if (vbuf->bytesused <= net_buf_tailroom(buf)) {
+		/* Very short video buffer fitting in the first packet */
+		*next_vbuf_offset = vbuf->bytesused;
+	} else {
+		/* Pad the USB buffer until the next video buffer pointer is aligned for UDC */
+		while (!IS_UDC_ALIGNED((uintptr_t)&vbuf->buffer[net_buf_tailroom(buf)])) {
+			net_buf_add_u8(buf, 0);
+			((struct uvc_payload_header *)buf->data)->bHeaderLength++;
+		}
+
+		*next_vbuf_offset = net_buf_tailroom(buf);
+	}
+
+	net_buf_add_mem(buf, vbuf->buffer, *next_vbuf_offset);
+
+	/* If this new USB transfer will complete this frame */
+	if (fmt->pitch == 0 || *next_line_offset >= fmt->height) {
+		LOG_DBG("Last USB transfer for this buffer");
+
+		/* Flag that this current transfer is the last */
+		((struct uvc_payload_header *)buf->data)->bmHeaderInfo |=
+			UVC_BMHEADERINFO_END_OF_FRAME;
+
+		/* Toggle the Frame ID of the next vbuf */
+		data->payload_header.bmHeaderInfo ^= UVC_BMHEADERINFO_FRAMEID;
+
+		*next_line_offset = 0;
+	}
+
+	return buf;
+}
+
+/*
+ * Handling the continuation of USB transfers marked by 'v' below:
+ *              v                                       v
+ * [hdr:data:::][data::::::::::::::::::::] [hdr:data:::][data::::::::::::::::::::] ...
+ *      [vbuf::::::::::::::::::::::::::::]      [vbuf::::::::::::::::::::::::::::] ...
+ */
+static struct net_buf *uvc_continue_transfer(const struct device *dev,
+					     struct video_buffer *const vbuf,
+					     size_t *const next_line_offset,
+					     size_t *const next_vbuf_offset)
+{
+	struct uvc_data *data = dev->data;
+	struct video_format *fmt = &data->video_fmt;
+	struct net_buf *buf;
+	/* Workaround net_buf that uses uint16_t storage for lengths and offsets */
+	const size_t max_len = 0xf000;
+	const size_t buf_len = MIN(max_len, vbuf->bytesused - data->vbuf_offset);
+
+	/* Directly pass the vbuf content with zero-copy */
+	buf = net_buf_alloc_with_data(&uvc_buf_pool, vbuf->buffer + data->vbuf_offset,
+				      buf_len, K_NO_WAIT);
+	if (buf == NULL) {
+		LOG_DBG("Cannot allocate continuation USB buffer for now");
+		return NULL;
+	}
+
+	/* If uncompressed and line-based format, update the next line position in the frame */
+	if (fmt->pitch > 0) {
+		*next_line_offset = vbuf->line_offset + buf->len / fmt->pitch;
+	}
+
+	/* The entire video buffer is now submitted */
+	*next_vbuf_offset = data->vbuf_offset + buf_len;
+
+	return buf;
+}
+
+static int uvc_reset_transfer(const struct device *dev)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	struct uvc_buf_info *bi;
+	struct net_buf *buf;
+	int ret;
+
+	LOG_DBG("Stream restarted, terminating the transfer after %u bytes", data->vbuf_offset);
+
+	buf = net_buf_alloc_with_data(&uvc_buf_pool, "", 1, K_NO_WAIT);
+	if (buf == NULL) {
+		LOG_DBG("Cannot allocate ZLP USB buffer for now");
+		return -ENOMEM;
+	}
```

Comment:
This is also likely to result in an unaligned buffer being submitted. I observed this issue on an nRF54LM20 device where DWC2 driver complained about an unaligned buffer after a stream restart. I traced the issue down to this point. You can replace it with
```c
	`buf = net_buf_alloc_len(&uvc_buf_pool, 0, K_NO_WAIT);`
	if (buf == NULL) {
```
All the UDC driver should be able to handle it. (The bi->zlp flag is where you have transfer length multiple of MPS but < length host requested, e.g. in the control transfers and some other cases).